crates/bootstrap_client/src/lib.rs

```diff
@@ -4,52 +4,246 @@
+        struct AuthToken {
+            auth_token: String,
+        }
+
+        let auth_token: AuthToken = serde_json::from_str(&token)
+            .map_err(|err| K2Error::other_src("Authenticate Failed", err))?;
+
+        *self.auth_token.lock().unwrap() = Some(auth_token.auth_token);
+
+        Ok(())
+    }
+}
+
+enum Res<T> {
+    Ok(T),
+    Auth,
+    Err(K2Error),
+}
+
+impl<T> Res<T> {
+    fn needs_auth(&self) -> bool {
+        matches!(self, Self::Auth)
+    }
+}
+
+impl<T> From<std::result::Result<T, ureq::Error>> for Res<T> {
+    fn from(r: std::result::Result<T, ureq::Error>) -> Self {
+        match r {
+            Ok(t) => Self::Ok(t),
+            Err(ureq::Error::Status(401, _)) => Self::Auth,
+            Err(err) => Self::Err(K2Error::other(err)),
+        }
+    }
+}
+
+impl<T> From<std::io::Result<T>> for Res<T> {
+    fn from(r: std::io::Result<T>) -> Self {
+        match r {
+            Ok(t) => Self::Ok(t),
+            Err(err) => Self::Err(K2Error::other(err)),
+        }
+    }
+}
+
+impl<T> From<Res<T>> for K2Result<T> {
+    fn from(r: Res<T>) -> Self {
+        match r {
+            Res::Ok(t) => Ok(t),
+            Res::Auth => Err(K2Error::other("Unauthorized")),
+            Res::Err(err) => Err(err),
+        }
+    }
+}
+
 /// Send the agent info, for the given space, to the bootstrap server.
 ///
-/// Note the `blocking_` prefix. This is a hint to the caller that if the function is used in
-/// an async context, it should be treated as a blocking operation.
+/// Note the `blocking_` prefix. This is a hint to the caller that if the
+/// function is used in an async context, it should be treated as a blocking
+/// operation.
 pub fn blocking_put(
+    server_url: Url,
+    agent_info: &AgentInfoSigned,
+) -> K2Result<()> {
+    blocking_put_auth(server_url, agent_info, None)
+}
+
+/// Send the agent info, for the given space, to the bootstrap server.
+///
+/// Note the `blocking_` prefix. This is a hint to the caller that if the
+/// function is used in an async context, it should be treated as a blocking
+/// operation.
+pub fn blocking_put_auth(
     mut server_url: Url,
     agent_info: &AgentInfoSigned,
+    auth_material: Option<&AuthMaterial>,
 ) -> K2Result<()> {
+    server_url.set_path("authenticate");
+    let auth_url = server_url.as_str().to_string();
+
     server_url.set_path(&format!(
         "bootstrap/{}/{}",
         base64::prelude::BASE64_URL_SAFE_NO_PAD.encode(&**agent_info.space),
         base64::prelude::BASE64_URL_SAFE_NO_PAD.encode(&**agent_info.agent),
     ));
+    let put_url = server_url.as_str().to_string();
+
+    if let Some(auth_material) = &auth_material {
+        auth_material.priv_authenticate(&auth_url, AuthType::IfUninit)?;
+    }
 
     let encoded = agent_info.encode()?;
-    ureq::put(server_url.as_str())
-        .send_string(&encoded)
-        .map_err(|e| K2Error::other_src("Failed to put agent info", e))?;
 
-    Ok(())
+    fn priv_put(
```

Comment:
Extract a module-private function?

Reply:
Is there a reason to extract it? It's only useful inside this function, and only needs to be a function itself to DRY out the retry in case of initial token failure.

Reply:
Fair enough, I just prefer shorter functions but I get that it's scoped to this function :) There's no reason beyond that no